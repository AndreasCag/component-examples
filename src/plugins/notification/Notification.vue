<script>
import classNames from 'classnames';
import CrossIcon from './CrossIcon.vue';

export default {

  props: {
    data: {
      type: Object,
      required: true,
    },
  },

  computed: {

    renderedContent() {
      const { data } = this;

      if (data.render) {
        return data.render(this.close);
      }

      return [
        <div class="header">
          {data.header}
        </div>,
        <div class="content">
          {data.content}
        </div>,
        <CrossIcon
          class="cross-icon"
          onClick={this.close}
        />,
      ];
    },

    notificationClass() {
      switch (this.data.mode) {
        case 'error':
          return 'notification_error';
        case 'warning':
        default:
          return 'notification_warning';
      }
    },

  },

  methods: {

    close() {
      this.$emit('close');
    },

  },

  render() {
    const { renderedContent, notificationClass } = this;

    return (
      <div class={classNames('notification', notificationClass)}>
        {renderedContent}
      </div>
    );
  },

};
</script>

<style scoped>
.notification {
  position: relative;
  padding: 16px 32px;
  box-shadow: 0px 3px 5px rgba(0, 0, 0, .5);
  color: white;
  font-size: 16px;
  line-height: 22px;
  border-radius: 2px;
}

.notification_warning {
  background-color: rgba(189, 54, 47, .8)
}

.header {
  font-weight: bold;
  margin-bottom: 8px;
  margin-right: 4px;
}

.cross-icon {
  position: absolute;
  right: 16px;
  top: 16px;
  width: 16px;
  height: 16px;
  fill: white;
  cursor: pointer;
}
</style>
