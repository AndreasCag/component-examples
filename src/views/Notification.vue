<template>
  <div>
    Notification
    <button @click="addNotification">
      Add notification
    </button>
  </div>
</template>

<script>
export default {

  methods: {

    addNotification() {
      this.$addNotification({
        header: 'Test Test Test Test Test Test Test Test Test ',
        content: 'Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text Text ',
        duration: 40000000,
      });
    },

  },

};
</script>
