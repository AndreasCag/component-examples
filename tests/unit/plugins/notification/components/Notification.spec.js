import Notification from '@/plugins/notification/Notification.vue';
import CrossIcon from '@/plugins/notification/CrossIcon.vue';
import { shallowMount, mount } from '@vue/test-utils';
import Vue from 'vue';

// H variable is used in JSX
// eslint-disable-next-line no-unused-vars
const h = new Vue().$createElement;

describe('Notification.vue', () => {
  it('renders header and content', () => {
    const data = {
      header: 'Test header',
      content: 'Test content',
    };

    const notificationWrapper = shallowMount(Notification, {
      propsData: { data },
    });

    const notificationText = notificationWrapper.text();

    expect(notificationText).toMatch(data.header);
    expect(notificationText).toMatch(data.content);
  });

  it('adds notification_warning class to message without mode', () => {
    const notificationWrapper = shallowMount(Notification, {
      propsData: { data: {} },
    });

    expect(notificationWrapper.classes()).toContain('notification_warning');
  });

  it('adds notification_warning class to message with warning mode', () => {
    const data = {
      mode: 'warning',
    };

    const notificationWrapper = shallowMount(Notification, {
      propsData: { data },
    });

    expect(notificationWrapper.classes()).toContain('notification_warning');
  });

  it('adds notification_error class to message with error mode', () => {
    const data = {
      mode: 'error',
    };

    const notificationWrapper = shallowMount(Notification, {
      propsData: { data },
    });

    expect(notificationWrapper.classes()).toContain('notification_error');
  });

  it('emits close event', () => {
    const notificationWrapper = mount(Notification, {
      propsData: { data: {} },
    });

    notificationWrapper.find(CrossIcon).trigger('click');

    expect(notificationWrapper.emitted().close.length).toBe(1);
  });

  it('renders passed content', () => {
    const data = {
      render: () => [
        <div class="test-header">
          test header
        </div>,
        <div class="test-subheader">
          test subheader
        </div>,
        <div class="test-content">
          test content
        </div>,
      ],
    };

    const notificationWrapper = shallowMount(Notification, {
      propsData: { data },
    });

    expect(notificationWrapper.element).toMatchSnapshot();
  });

  it('passes close function as argument in render function', () => {
    const mockRenderFunction = jest.fn(() => (
      <div>
        test data
      </div>
    ));

    const data = {
      render: mockRenderFunction,
    };

    const notificationWrapper = shallowMount(Notification, {
      propsData: { data },
    });

    expect(mockRenderFunction).toBeCalledWith(expect.any(Function));
    expect(notificationWrapper.emitted().close).toBeFalsy();

    // call close function
    mockRenderFunction.mock.calls[0][0]();

    expect(notificationWrapper.emitted().close.length).toBe(1);
  });
});
