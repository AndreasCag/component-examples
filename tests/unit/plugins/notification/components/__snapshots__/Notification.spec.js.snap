// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification.vue renders passed content 1`] = `
<div
  class="notification notification_warning"
>
  <div
    class="test-header"
  >
    test header
  </div>
  <div
    class="test-subheader"
  >
    test subheader
  </div>
  <div
    class="test-content"
  >
    test content
  </div>
</div>
`;
